### ISSUE TITLE
Min for nullable types

### ISSUE THREAD

**Issue by StephenCleary** [dotnet/reactive#869]
`Min` doesn't treat `null` as a "value" for nullable types. I was expecting `Min` to produce `null` if a `null` is present in the input observable.

This behavior is notably different than System.Interactive, where `Min` does respect and return `null` if present.

**Comment by quinmars** [comment]
Looking at the code, `Min` should return `null` for an empty or only `null` containing sequence. Are you expecting something else? What type are you using? There are some specialist versions for the common numeric values.

**Comment by quinmars** [comment]
Oh, and are you talking about Rx or AsyncIx? My answer was for Rx.

**Comment by StephenCleary** [comment]
> Are you expecting something else?

Yes, though I also see the validity of the "special null" semantic.

What I expected: The default comparer for nullable numeric types like `int?` treats `null` as "less than" any other value, so given a `null` input value, I'd expect the following:

```C#
var values = new int?[] { int.MaxValue, null, int.MinValue }.ToObservable();
var result = await values.Min().LastAsync();
Assert.Null(result); // fails; `result` is `int.MinValue`
```

This behavior is different than Ix, which has the following behavior:

```C#
var values = new int?[] { int.MaxValue, null, int.MinValue };
var result = values.Min();
Assert.Null(result); // succeeds; `result` is `null`
```

The current semantics assume `null` is not a "real" value for sequences of nullable numeric types. I can see the validity of that argument; but to me it is more natural to treat `null` as a "real" value for `IObservable<int?>`, as it is for most other operators, and to use `Where` to filter out `null`s if desired.

**Comment by quinmars** [comment]
Well, it is the behaviour like in Linq:

```chsarp
using System;
using System.Linq;

public static class Program {
    public static void Main() {
        var result = new int?[] { int.MaxValue, null, int.MinValue }.Min();
        Console.WriteLine(result);                // result == -2147483648
    }
}
```
See https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQQHYAsoYMYQAmqcyAlhgOYA0MICcGAPgAIAMABCwIwDcAsACh2XbgDoAMhQCOA4QCY2QlgGZRANi7zRAdg4BvIR2Nc1PTSwAsHALJQKACgCUBoyfcA3KAg5JkcABsYDgBeDgwIAHcOChgAfgBtAF0DGIwYMTsADwA1KAC4CGpwwIDi2MyKPIKIDgBfSoxnOXd3HgBOBz9AmCcWkzqhOqA===

Afaik, the behaviour was changed from Rx1 to Rx2 to reflect the `IEnumerable<T>` semantics. Also the [documentation of `Enumerable.Min()`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.min?view=netframework-4.7.2) gives the following example:

```csharp
int?[] grades = { 78, 92, null, 99, 37, 81 };

int? min = grades.Min();

Console.WriteLine("The lowest grade is {0}.", min);

/*
 This code produces the following output:

 The lowest grade is 37.
*/
```

**Comment by StephenCleary** [comment]
OK. In that case, Ix should probably be changed to match. My example code above is incorrect; the correct code should be:

```C#
var values = new int?[] { int.MaxValue, null, int.MinValue };
var result = values.Min(Comparer<int?>.Default);
Assert.Null(result); // succeeds; `result` is `null`
```

which uses the actual Ix `Min` and not the LINQ one.

**Comment by onovotny** [comment]
@StephenCleary Not sure I follow completely, is there a bug in Ix we need to address here?

**Comment by StephenCleary** [comment]
It's an inconsistency. All the following tests pass:

```C#
[Fact]
public async Task RxMin()
{
    var values = new int?[] { 0, null }.ToObservable();
    var result = await values.Min().LastAsync();
    Assert.Equal(0, result);
}

[Fact]
public async Task RxMin_WithComparer()
{
    var values = new int?[] { 0, null }.ToObservable();
    var result = await values.Min(Comparer<int?>.Default).LastAsync();
    Assert.Equal(0, result);
}

[Fact]
public void LinqMin()
{
    var values = new int?[] { 0, null };
    var result = values.Min();
    Assert.Equal(0, result);
}

[Fact]
public void IxMin()
{
    var values = new int?[] { 0, null };
    var result = values.Min(Comparer<int?>.Default);
    Assert.Null(result);
}
```

Not sure if it's worth fixing. I have no use case; I just ran into this edge case when testing my [comparers library](https://github.com/StephenCleary/Comparers).

**Comment by quinmars** [comment]
"Fixing" this would be a breaking change. In order to be consistent, the implementation of `MinBy<T>()` would also need to be reconsidered. But what should be the return value of `new int?[] {}.MinBy(x => x)`? An empty list or a list of a single `null`?

However, it would be possible to add a new method that match the Linq behavior:

```csharp
public T Min(this IEnumerable<T> source, IComparer<T> comparer, bool treatNullSpecial)
{
    return (treatNullSpecail && default(T) == null)
        ? ...
        : source.Min(comparer);
}
```

**Comment by bartdesmet** [comment]
Yes, this is a long-standing inconsistency that dates back all the way to Rx v1.0. It fell through the cracks in the final beta days where we were noodling around with the return type for aggregates and at some point `null` had a special meaning. By the time we shipped v1.0, the inconsistent nullable behavior stuck and it was reconsidered for "fixing" in v2.0, but considered too risky given the breaking change potential, with the advice to use a manual `Aggregate` (though verbose and clunky) if the LINQ to Objects `null` behavior is desired.

At this point, it's established (though inconsistent and very unfortunate) behavior, and the risk of breaking users outweighs the desire for consistency. Note for `System.Linq.Async`, the behavior is consistent with synchronous LINQ to Objects.
